src/lib.rs

```diff
@@ -0,0 +1,459 @@
+    type NFTDepositPerByte: Get<BalanceOf<Self>>;
+
+    /// The amount of balance that must be deposited per validation function registry.
+    type NFTValidationRegistryDeposit: Get<BalanceOf<Self>>;
+
+    /// Currency type for this module.
+    type Currency: ReservableCurrency<Self::AccountId>
+        + LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
+}
+
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        /// Validation function not registered
+        ValidationFnNotExisted,
+
+        /// Uid not registered
+        UidNotRegistered,
+
+        /// Token not minted
+        TokenNotMinted,
+
+        /// Token already existed
+        TokenAlreadyExisted,
+
+        // Now token owner
+        WrongTokenOwner,
+
+        // Metadata length invalid
+        MetadataLengthInvalid,
+
+        // Not validation function
+        NotValidationFunction,
+
+        // Proof validation failed
+        ProofValidationFailure,
+
+        // Proof validation failed
+        DocumentNotAnchored,
+
+        // Token URI length
+        // TokenURILengthInvalid,
+    }
+}
+
+decl_event!(
+    pub enum Event<T> 
+        where
+        <T as system::Trait>::AccountId,
+        <T as system::Trait>::Hash {
+        // Account register a new Uid with smart contract
+        NewRegistry(AccountId, RegistryUid),
+        
+        // New NFT created
+        MintNft(RegistryUid, Hash),
+    }
+);
+
+decl_storage! {
+    trait Store for Module<T: Trait> as NftRegistry {
+        // Each registry uid include a validation function in smart contract
+        pub ValidationFn get(validator_of): map hasher(blake2_256) RegistryUid => Option<T::AccountId>;
+
+        // Next Registry id
+        pub NextRegistryId: RegistryUid;
+
+        // Registry Metadata
+        pub TokenByRegistryId get(token_by_registry_id): map T::Hash => RegistryUid;
+
+        // Metadata for each token id
+        pub RegistryTokenMetadata get(registry_token_metadata): map T::Hash => Vec<u8>;
+
+        // Metadata min length
+        pub MinMetadataLength get(min_token_metadata_length) config(): u32;
+
+        // Metadata max length
+        pub MaxMetadataLength get(max_token_metadata_length) config(): u32;
+
+        // Token URI not needed yet.
+        // Token's URI
+        // pub RegistryTokenURI get(registry_token_uri): map hasher(blake2_256) T::Hash => Vec<u8>;
+        // Token URI min length
+        // pub MinTokenURILength get(min_token_uri_length) config(): u32;
+        // Token URI max length
+        // pub MaxTokenURILength get(max_token_uri_length) config(): u32;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin  {
+        fn deposit_event() = default;
+
+        // Fee for store one byte
+        const NFTDepositPerByte: BalanceOf<T> = T::NFTDepositPerByte::get();
+
+        // Fee for one NFT metadata storage
+        const NFTDepositBase: BalanceOf<T> = T::NFTDepositBase::get();
```

Comment:
1. Deposit change will make the runtime much more complidated. We don't know how much we need unreserve for burned token. Or we need record the reserved currency for each token. 
2. I should use constants instead of get from trait in the following method in compute_metadata_fee.

Reply:
1. We definitely need, deposits will change. See my comment further down. 
2. Can you elaborate why we need to use constants instead of getting it from thetrait?

Reply:
If we store the deposit, then we don't need define them as constants. I will remove the variables.